File: java_implementation/src/main/java/com/example/taskapi/application/usecases/CompleteTaskUseCase.java
package com.example.taskapi.application.usecases;

import com.example.taskapi.domain.enums.TaskStatus;
import com.example.taskapi.domain.model.Task;
import com.example.taskapi.domain.repository.TaskRepositoryPort;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CompleteTaskUseCase {
    private final TaskRepositoryPort taskRepository;

    public Optional<Task> execute(UUID taskId) {
        Optional<Task> taskOptional = taskRepository.getById(taskId);
        if (taskOptional.isEmpty()) {
            return Optional.empty();
        }

        Task task = taskOptional.get();
        if (task.getStatus() == TaskStatus.DONE) {
            return Optional.of(task);
        }

        task.setStatus(TaskStatus.DONE);
        return Optional.of(taskRepository.update(task));
    }
}

File: python_implementation/src/application/use_cases/tasks/complete_task_use_case.py
from typing import Optional
from uuid import UUID
from src.domain.abstractions.repositories.abstract_task_repository import AbstractTaskRepository
from src.domain.entities.task_entity import TaskEntity
from src.domain.enums.task_enums import TaskStatus


class CompleteTaskUseCase:
    """Use case for marking a task as complete"""
    
    def __init__(self, task_repository: AbstractTaskRepository):
        self.task_repository = task_repository
    
    def execute(self, task_id: UUID) -> Optional[TaskEntity]:
        task = self.task_repository.get_by_id(task_id)
        if not task:
            return None
        
        if task.status == TaskStatus.DONE:
            return task
        
        # Update status to DONE
        task.status = TaskStatus.DONE
        return self.task_repository.update(task)
